Speedier user organizations

sandersonet opened the pull request:
This PR makes two minor changes that net pretty significant speed improvements for Gridiron (and probably dashboard in general).

First, `User#organizations` was loading `role.organization` for every single role, even if the organization was already side loaded from another role.  This PR improves `User#organizations` to establish uniqueness before making any requests.  This removes an AUTH request for each of the user's roles--In the case of gridiron, that's a lot of roles for the initial user!

Secondly,  Auth has been embedding users in membership responses for a while.  Since `Role#users` just maps over memberships to return users, we can again remove several AUTH requests by just updating from `belongs_to` to `embeds_one` for `Membership#user`. 

---

cc @fancyremarker @krallin

krallin commented on lib/aptible/auth/user.rb:
```diff
@@ -14,7 +14,9 @@ class User < Resource
-        roles.map(&:organization).uniq(&:id)
+        # Establish uniqueness of requests before loading all organizations
+        # We can do this by reading the `organization` link for each role
+        roles.map(&:links).map(&:organization).uniq(&:base_href).map(&:get)
```
Does this not work properly if you use `href`?

The reason I ask is because the underlying library we use (HyperResource) has some support for templated URLs, and `base_href` would return the template, whereas the actual HREF that would be used for a `get` is present at `href`.

Since we don't currently use those templates, I'm fairly certain `href` should just passthrough to `base_href`:

```
  def href
    if self.templated?
      filtered_params = self.parent_resource.outgoing_uri_filter(params)
      URITemplate.new(self.base_href).expand(filtered_params)
    else
      self.base_href
    end
  end
```

But, assuming we can use this, using `href` is probably more robust.

krallin commented on lib/aptible/auth/membership.rb:
```diff
@@ -2,7 +2,7 @@ module Aptible
     class Membership < Resource
       belongs_to :role
-      belongs_to :user
+      embeds_one :user
```
Using this actually subtly breaks something (although you're unlikely to notice). Unfortunately, it's due to a bug in HyperResource / aptible-resource (and arguably in auth.aptible.com), which we haven't gotten around to fixing yet.

Basically, auth.aptible.com does *not* decorate the records it passes when rendering embeds (or at the very least, it doesn't do so consistently, and it doesn't do it in this case). As a result, the `_type` field in the embed will be empty

When it reads the embed, in the absence of a `_type` field, HyperResource will assumes that the class is unchanged. As a result, with this change, `role.users` will return `Aptible::Auth::Membership` instances that actually represent Users (!).

You'll be able to access the fields on the result, but:

- Type coercion won't work. For example, since `Aptible::Auth::User` includes `field :updated_at, type: Time
`, but `Aptible::Auth::Membership` does not, you won't get type coercion for `updated_at`.
- All the methods defined `Aptible::Auth::User` will be missing. For example, calling `#organizations` will throw an error, because you're calling it on a `Aptible::Auth::Membership` but it's defined for `Aptible::Auth::User`.

Unfortunately, we can't just systematically decorate the records right now (I wish we could), because then HyperResource will throw an exception whenever it encounters a class it does not know about, which means that embedding any new record type will break old clients that don't know about it.

Notably, this happened when we rolled out Database Credentials: we embed them in the DB object, so if API submits a `_type`, clients that don't know about Database Credentials crash when they try to load a Database that embeds them.

----

All of this to say that this change might have unintended side effects.

The right fix here is of course to rethink our client library implementation and fix these problems there (we'll need the new client to send some opt-in `Accept` header to communicate that it expects everything to be decorated and knows how to handle record types it does not know about).

That being said, I think all clients that know about Memberships should know about users as well, so assuming you want to get this out the door before we rewrite said client (!), could you add decorate the Users embedded in Memberships in `auth.aptible.com`?

Thanks!

sandersonet commented:
@krallin - Thank you for the thorough review and feedback.  I've updated auth to decorate embedded users and updated this PR to reference `:href` rather than `:base_href`. We can wait to merge and release this until the Auth PR is shipped.

krallin commented:
Thanks!
